cpp/src/io/parquet/reader_impl_chunking_utils.cuh

```diff
@@ -0,0 +1,640 @@
+  {
+    return chunk.compressed_size;
+  }
+};
+
+/**
+ * @brief Functor which returns the number of rows in a page for a flat column
+ */
+struct flat_column_num_rows {
+  ColumnChunkDesc const* chunks;
+  __device__ inline size_type operator()(PageInfo const& page) const
+  {
+    // Ignore dictionary pages and pages belonging to any column containing repetition (lists)
+    if ((page.flags & PAGEINFO_FLAGS_DICTIONARY) ||
+        (chunks[page.chunk_idx].max_level[level_type::REPETITION] > 0)) {
+      return 0;
+    }
+    return page.num_rows;
+  }
+};
+
+/**
+ * @brief Stores basic information about pages compressed with a specific codec
+ */
+struct codec_stats {
+  Compression compression_type  = Compression::UNCOMPRESSED;
+  size_t num_pages              = 0;
+  int32_t max_decompressed_size = 0;
+  size_t total_decomp_size      = 0;
+
+  enum class page_selection { DICT_PAGES, NON_DICT_PAGES };
+
+  void add_pages(host_span<ColumnChunkDesc const> chunks,
+                 host_span<PageInfo> pages,
+                 page_selection selection);
+};
+
+/**
+ * @brief Functor which retrieves per-page decompression information.
+ */
+struct get_decomp_info {
+  device_span<ColumnChunkDesc const> chunks;
+  __device__ inline decompression_info operator()(PageInfo const& p) const
+  {
+    return {parquet_compression_support(chunks[p.chunk_idx].codec).first,
+            1,
+            static_cast<size_t>(p.uncompressed_page_size),
+            static_cast<size_t>(p.uncompressed_page_size)};
+  }
+};
+
+/**
+ * @brief Functor which accumulates per-page decompression information.
+ */
+struct decomp_sum {
+  __device__ inline decompression_info operator()(decompression_info const& a,
+                                                  decompression_info const& b) const
+  {
+    return {a.type,
+            a.num_pages + b.num_pages,
+            cuda::std::max(a.max_page_decompressed_size, b.max_page_decompressed_size),
+            a.total_decompressed_size + b.total_decompressed_size};
+  }
+};
+
+/**
+ * @brief Functor which check if row count is not zero
+ */
+struct row_counts_nonzero {
+  __device__ inline bool operator()(size_type count) const { return count > 0; }
+};
+
+/**
+ * @brief Functor which checks if row count is non zero and different from expected value
+ */
+struct row_counts_different {
+  size_type const expected;
+  __device__ inline bool operator()(size_type count) const
+  {
+    return (count != 0) && (count != expected);
+  }
+};
+
+/**
+ * @brief Functor which computes the total data size for a given type of cudf column
+ *
+ * In the case of strings, the return size does not include the chars themselves. That
+ * information is tracked separately (see PageInfo::str_bytes).
+ */
+struct row_size_functor {
+  __device__ inline size_t validity_size(size_t num_rows, bool nullable)
+  {
+    return nullable ? cudf::num_bitmask_words(num_rows) * sizeof(bitmask_type) : 0;
```

Comment:
Using `cudf::num_bitmask_words` here instead of manual calculation with magic numbers

Reply:
This can be done in a follow-on PR. I don't want to hold up this one or add more code than necessary here.

Reply:
>This can be done in a follow-on PR. I don't want to hold up this one or add more code than necessary here.

Yep, that was my motivation behind the current changes.